internal/juju/applications.go

```diff
@@ -410,6 +410,13 @@ func (c applicationsClient) deployFromRepository(applicationAPIClient Applicatio
 		return errors.Join(errs...)
 	}
 
+	// publishing to a closed channel automatically upgrade the channel. And this could
+	// lead to an inconsistent terraform state. This check is to error out to the user,
+	// so they can purposely select the track.
+	if deployInfo.Channel != transformedInput.charmChannel {
+		return fmt.Errorf("you are trying to publish to a closed channel. Check the channel for the %s", transformedInput.applicationName)
```

Comment:
You're failing after the application has been deployed, but no local OCI image resources are uploaded? Thus the terraform state will not be saved, but the application exists in juju.

This sounds like a worse state to be in than the one you're trying to solve.